src/job_cache/daemon_cache.cpp

```diff
@@ -0,0 +1,1052 @@
+    evict_stdout = stdoutPipe[read_side];
+  }
+}
+
+void DaemonCache::reap_evict_loop() {
+  close(evict_stdin);
+  close(evict_stdout);
+  waitpid(evict_pid, nullptr, 0);
+}
+
+DaemonCache::~DaemonCache() { reap_evict_loop(); }
+
+void DaemonCache::handle_new_client() {
+  // Accept the new client socket. Because of needing to perform
+  // multiple read's per client IO event, we have to make the client
+  // socket use non-blocking reads. This allows us to consume multiple
+  // read calls without blocking the other clients if one of them waits.
+  int accept_fd = accept4(listen_socket_fd, nullptr, nullptr, SOCK_CLOEXEC | SOCK_NONBLOCK);
+  if (accept_fd == -1) {
+    log_fatal("accept(%s): %s", key.c_str(), strerror(errno));
+  }
+
+  poll.add(accept_fd);
+  message_parsers.insert({accept_fd, MessageParser(accept_fd)});
+  log_info("new client connected: %d", accept_fd);
+}
+
+void DaemonCache::handle_msg(int client_fd) {
+  // In case multiple read events have been enqueued since the
+  // last epoll_wait, we have to perform all the reads that
+  // have been enqueued.
+  MessageParserState state;
+  std::vector<std::string> msgs;
+
+  auto it = message_parsers.find(client_fd);
+  if (it == message_parsers.end()) {
+    log_fatal("unreachable: message_parsers out of sync with poll. client_fd = %d", client_fd);
+  }
+
+  state = it->second.read_messages(msgs);
+
+  for (const auto &msg : msgs) {
+    log_info("msg: %s", msg.c_str());
+
+    JAST json;
+    std::stringstream parseErrors;
+    if (!JAST::parse(msg, parseErrors, json)) {
+      log_fatal("DaemonCache::handle_msg(): failed to parse client request");
+    }
+
+    if (json.get("method").value == "cache/read") {
+      FindJobRequest req(json.get("params"));
+      auto match_opt = read(req);
+
+      JAST res(JSON_OBJECT);
+      res.add("method", "cache/read");
+
+      if (match_opt) {
+        res.add("params", match_opt->to_json());
+      } else {
+        res.add("params", JSON_NULLVAL);
+      }
+
+      send_json_message(client_fd, res);
+    }
+
+    if (json.get("method").value == "cache/add") {
+      AddJobRequest req(json.get("params"));
+      add(req);
+    }
+  }
+
+  // If the file was closed, remove from epoll and close it.
+  if (state == MessageParserState::StopSuccess) {
+    log_info("closing client fd = %d", client_fd);
+    poll.remove(client_fd);
+    close(client_fd);
+    message_parsers.erase(client_fd);
+    if (message_parsers.empty()) {
+      exit_now = true;
+      log_info("All clients disconnected, exiting.");
+    }
+    return;
+  }
+
+  // If there's an error, consider which one it is or fail
+  if (state == MessageParserState::StopFail) {
+    // These sockets are non-blocking because we have to try reading
+    // them multiple times. This means that often we'll wind up
+    // attempting N+1 reads and one of them will fail like this
+    // without closing.
+    if (errno == EAGAIN || errno == EWOULDBLOCK) return;
```

Comment:
I think this code migrated here during a refactor but should be in handle_msg or not exist at all. I don't remember the context in which I wrote this error message so I'm not sure. I'll take a look a bit later to see what I meant by this.